### PR TITLE
Improve type tracking through recipe.updateToClone.

### PR DESCRIPTION
**Pull request by shans**
This used to be typed as returning any; now it returns the same type
it is given, which means the output dict members all automatically
have the correct type.

It does make generic uses slightly more painful (e.g. see
convert-constraints-to-connections.ts) but even here we've now got
more type safety as we're forced to declare what we expect each output
to be.

### CODE REVIEW

**Review by mykmartin — APPROVED** [review]

File: src/planning/strategies/convert-constraints-to-connections.ts [hunk 1]
```diff
@@ -194,9 +194,9 @@ export class ConvertConstraintsToConnections extends Strategy {
                 let recipeHandleConnection = recipeParticle.connections[connection];
                 if (recipeHandleConnection == undefined) {
                   recipeHandleConnection =
-                      recipeParticle.addConnectionName(connection);
+                      recipeParticle.addConnectionName(connection) as HandleConnection;
```
Comment (mmandlis) [inline src/planning/strategies/convert-constraints-to-connections.ts:197]: nit - why is this needed, `addConnectionName` already returns a `HandleConnection`?

Comment (shans) [reply src/planning/strategies/convert-constraints-to-connections.ts:197]: I think I added that first (when the type of recipeParticle wasn't known) but you're right, it isn't necessary now. Will remove in a follow-up PR

Comment (shans) [reply src/planning/strategies/convert-constraints-to-connections.ts:197]: https://github.com/PolymerLabs/arcs/pull/6135
